### PR TITLE
Fix modifySeparatedNodeList

### LINKED ISSUE #25273: Bug in TreeModifier.modifySeparatedNodeList method
**Description:**
https://github.com/ballerina-platform/ballerina-lang/blob/c4f3e100fa1cbd4f003f689ecf90fe560d63689b/compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/TreeModifier.java#L3100-L3101

* Here in L3101 we have called `modifyGenericNodeList()` which only iterates through the alternative elements as we are parsing a `SeparatedNodeList` object, not  `NodeList` object.

* So when there's no modification, that method simply returns the original SeparatedNodeList which is fine. 
But when there's a modification, it returns a NodeList object in which there are no separators, as no processing has been done for separators.

* Subsequently, when we are formatting with new formatter and when there's a modification, separators just get vanished. 

**Suggested Labels (optional):**
Component/SyntaxAPI, Type/Bug

**Suggested Assignees (optional):**
@sameerajayasoma

### PR DESCRIPTION
**Pull request by m36dot**
## Purpose
fix modifySeparatedNodeList returning NodeList instead of SeperatedNodeList when SeperatedNodeList is modified.

Fixes #25273

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by lochana-chathura** [comment]
@m36dot I think all changes to `TreeModifier.java` will be gone once we use the `treegen tool` as It is a generated class. 
You will have to modify https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-treegen/src/main/resources/external_tree_modifier_template.mustache for the change.

**Comment by lochana-chathura** [comment]
@m36dot I think all changes to TreeModifier.java will be gone once we use the treegen tool as It is a generated class.

You may have to modify https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-treegen/src/main/resources/external_tree_modifier_template.mustache for the change.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/TreeModifier.java [hunk 1]
```diff
@@ -3070,14 +3070,23 @@ protected Node transformSyntaxNode(Node node) {
...
+
         for (int index = 0; index < nodeList.size(); index++) {
             T oldNode = nodeList.get(index);
             T newNode = modifyNode(oldNode);
             if (oldNode != newNode) {
                 nodeModified = true;
             }
-            newSTNodes[index] = newNode.internalNode();
+
+            if (nodeList instanceof SeparatedNodeList && index < nodeList.size() - 1) {
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/TreeModifier.java:3082]: Can we use separate methods for modifying `NodeList` and `SeparatedNodeList`?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/TreeModifier.java [hunk 2]
```diff
@@ -3070,14 +3070,23 @@ protected Node transformSyntaxNode(Node node) {
...
+                newSTNodes[2 * index] = newNode.internalNode();
+                newSTNodes[(2 * index) + 1] = ((SeparatedNodeList<?>) nodeList).getSeparator(index).internalNode();
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/TreeModifier.java:3084]: Looks like the separator is not modified.
Shall we add a test case to check whether the separator is also modified? e.g: remove all whitespaces

File: compiler/ballerina-parser/src/test/resources/tree/separated_node_list_modify.bal [hunk 3]
```diff
@@ -0,0 +1 @@
+int a = [a,b,c];
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/resources/tree/separated_node_list_modify.bal:1]: missing EOF newline

**Review by SupunS — CHANGES_REQUESTED** [review]

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/TreeModifier.java (hunk 2 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/TreeModifier.java:3084]: alright

File: compiler/ballerina-parser/src/test/resources/tree/separated_node_list_modify.bal (hunk 3 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/test/resources/tree/separated_node_list_modify.bal:1]: ok

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/TreeModifier.java (hunk 1 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/TreeModifier.java:3082]: ok

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/TreeModifier.java (hunk 2 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/syntax/tree/TreeModifier.java:3084]: ok

File: compiler/ballerina-treegen/src/main/resources/external_tree_modifier_template.mustache [hunk 4]
```diff
@@ -69,7 +69,42 @@ public abstract class {{className}} extends {{superClassName}}<Node> {
...
+        STNode[] newSTNodes = new STNode[nodeList.internalListNode.size()];
+
+        for (int index = 0; index < nodeList.size(); index++) {
+            T oldNode = nodeList.get(index);
+            T newNode = modifyNode(oldNode);
+            if (oldNode != newNode) {
+                nodeModified = true;
+            }
+
+            if (index != nodeList.size() - 1) {
```
Comment (SupunS) [inline compiler/ballerina-treegen/src/main/resources/external_tree_modifier_template.mustache:87]: Inverting the condition would simplify the logic i think.

File: compiler/ballerina-treegen/src/main/resources/external_tree_modifier_template.mustache [hunk 5]
```diff
@@ -69,7 +69,42 @@ public abstract class {{className}} extends {{superClassName}}<Node> {
...
+                Token newSeperator = modifyToken(oldSeperator);
+
+                if (oldSeperator != newSeperator) {
+                    nodeModified = true;
+                }
+
+                newSTNodes[2 * index] = newNode.internalNode();
+                newSTNodes[(2 * index) + 1] = newSeperator.internalNode();
+            } else {
+                newSTNodes[2 * index] = newNode.internalNode();
```
Comment (SupunS) [inline compiler/ballerina-treegen/src/main/resources/external_tree_modifier_template.mustache:98]: This line is common to both conditions. can take it out.

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/tree/SyntaxTreeModifierTest.java [hunk 6]
```diff
@@ -109,6 +115,54 @@ public void testBinaryExprModification() {
...
+
+        ModuleVariableDeclarationNode newModuleVariableDeclarationNode =
+                (ModuleVariableDeclarationNode) newRoot.members().get(0);
+        ListConstructorExpressionNode newSeperatedlistNode =
+                (ListConstructorExpressionNode) newModuleVariableDeclarationNode.initializer();
+
+        Assert.assertEquals(oldSeperatedlistNode.expressions().separatorSize(),
+                newSeperatedlistNode.expressions().separatorSize());
+        Assert.assertEquals(oldSeperatedlistNode.expressions().size(),
+                newSeperatedlistNode.expressions().size());
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/tree/SyntaxTreeModifierTest.java:163]: These two asserts only checks for the counts.
We need to assert the content as well, to see whether they were updated correctly or not.

**Review by SupunS — CHANGES_REQUESTED** [review]

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/tree/SyntaxTreeModifierTest.java (hunk 6 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/tree/SyntaxTreeModifierTest.java:166]: ok

File: compiler/ballerina-parser/src/test/resources/tree/separated_node_list_modify_all_nodes_assert.bal [hunk 7]
```diff
@@ -0,0 +1 @@
+inta =[a  ,b  ,c];
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/resources/tree/separated_node_list_modify_all_nodes_assert.bal:1]: Shouldn't the WhiteSpaceMinutiaeRemover remove all the whitespaces?

**Review by SupunS — CHANGES_REQUESTED** [review]

File: compiler/ballerina-parser/src/test/resources/tree/separated_node_list_modify_all_nodes_assert.bal (hunk 7 above)
Comment (m36dot) [reply compiler/ballerina-parser/src/test/resources/tree/separated_node_list_modify_all_nodes_assert.bal:1]: it only removes the white space from all tokens not from nodes like SimpleNamedReferenceNode.

**Pull request merged**